Add OpenBSM service to collect socket events on macOS

Wajihulhassan opened the pull request:
Currently, Zeek-agent uses Endpoint Security framework service to collect file and process events on macOS; however, the Endpoint Security framework does not support socket events. Thus, to collect `connect` and `bind` socket events, this PR adds OpenBSM service. 

To use this service, the user only needs to add `nt` flag in `/etc/security/audit_control` configuration file of OpenBSM. After that socket events can be read from `/dev/auditpipe`. This PR takes care of reading events from that pipe and make socket events table. 

This OpenBSM service runs concurrently with Endpoint Security service which allows Zeek-agent to collect file, process, and socket events at the same time.

bbannier committed:
CMake: Add missing OpenSSL dependency to legacy Broker (#61)

`thirdparty_broker` introduces an implicit link dependency on OpenSSL's
libcrypto for any target linking against it. This means libcrypto needs
to be present or built before any of its dependents can be linked.

While even without the direct depedency this seems to have implicitly
happened most of the time when building with Make, before this patch
this did not work with the Ninja generator which detected before the
build that it did not have a rule to build libcrypto.

Adding the explicit link dependency in this patch fixes this issue.

Commit:
adding OpenBSM service to read socket events (connect and bind) in macos and make socket_events table

Wajihulhassan committed:
zeek-agent: Add file events table to macOS (#67)

bbannier commented on components/zeekopenbsm/src/openbsmconsumer.cpp:
```diff
@@ -0,0 +1,363 @@
+  }
+}
+
+void OpenbsmConsumer::extractReturn(IOpenbsmConsumer::Event &event,
+                                    tokenstr_t tok) {
+
+  switch (tok.id) {
+  case AUT_RETURN32: {
+    int error = 0;
+    if (au_bsm_to_errno(tok.tt.ret32.status, &error) == 0) {
+      if (error == 0) {
+        event.header.success = 1;
+      } else {
+        event.header.success = 0;
+      }
+    } else
+      event.header.success = 0;
+    break;
+  }
+  case AUT_RETURN64: {
+    int error = 0;
+    if (au_bsm_to_errno(tok.tt.ret64.err, &error) == 0) {
+      if (error == 0) {
+        event.header.success = 1;
+      } else {
+        event.header.success = 0;
+      }
+    } else
+      event.header.success = 0;
+    break;
+  }
+  }
+}
```
We could make this slightly less repetitive by extracting the status to expect in the switch,
```suggestion

void OpenbsmConsumer::extractReturn(IOpenbsmConsumer::Event &event,
                                    tokenstr_t tok) {
  u_char status = 0;
  switch (tok.id) {
  case AUT_RETURN32:
    status = tok.tt.ret32.status;
    break;
  case AUT_RETURN64:
    status = tok.tt.ret64.err;
    break;
  default:
    return;
  }

  int error = 0;
  if (au_bsm_to_errno(status, &error) == 0) {
    if (error == 0) {
      event.header.success = 1;
    } else {
      event.header.success = 0;
    }
  } else
    event.header.success = 0;
}

```

bbannier commented on components/zeekopenbsm/src/openbsmconsumer.cpp:
```diff
@@ -0,0 +1,363 @@
+      event.header.local_address = getIpFromToken(tok);
+      event.header.local_port = ntohs(tok.tt.sockinet_ex32.port);
+    } else {
+      event.header.remote_address = getIpFromToken(tok);
```
We could `assert(event.type == Event::Type::Connect)` here.

bbannier commented on components/zeekopenbsm/src/openbsmconsumer.cpp:
```diff
@@ -0,0 +1,363 @@
+  switch (tok.id) {
+  case AUT_SOCKINET32: {
+    if (event.type == Event::Type::Bind) {
+      event.header.remote_address = "0.0.0.0";
+      event.header.remote_port = 0;
```
Does it make sense to fill these fields with dummy values? Maybe we should just leave them unset (with the types you picked: empty string for `address`, `0` for `port`).

bbannier commented on components/zeekopenbsm/src/openbsmconsumer.cpp:
```diff
@@ -0,0 +1,363 @@
+      event.header.local_address = "0.0.0.0";
+      event.header.local_port = 0;
+    }
+    if (tok.tt.sockinet_ex32.family == 2) {
+      event.header.family = 2;
+    } else if (tok.tt.sockinet_ex32.family == 26) {
+      event.header.family = 10;
+    } else {
+      event.header.family = 0;
+    }
```
This is slightly hard to follow. Can you use some existing constants for the RHS?

bbannier commented on components/zeekopenbsm/src/openbsmconsumer.cpp:
```diff
@@ -0,0 +1,363 @@
+
+    break;
+  }
+  case AUT_SOCKINET128: {
```
See comments on `case AUT_SOCKINET32` above.

bbannier commented on components/zeekopenbsm/src/openbsmconsumer.cpp:
```diff
@@ -0,0 +1,363 @@
+  int reclen = au_read_rec(audit_pipe, &buffer);
+
+  if (reclen <= 0) {
+    return Status::failure("Could not openbsm fetch message");
```
Can we include `strerror(errno)` in this error message?

bbannier commented on components/zeekopenbsm/src/openbsmconsumer.cpp:
```diff
@@ -0,0 +1,363 @@
+
+  tokenstr_t tok;
+  std::vector<tokenstr_t> tokens{};
+  tokens.reserve(12);
```
Where is this number coming from? Shouldn't this be (derived from) `reclen` instead?

Wajihulhassan replied:
Basically, we need an upper bound on the number of token strings that can appear per socket bind/connect event. I could not find any documentation that can give me an upper bound value. Further, I don't think we can derive that number from `reclen` because `reclen` represents total bytes in all the token strings per event rather than the total number of token strings per event. So, I looked into osquery code and they were using 12 as an upper bound.

bbannier replied:
Let's either add a comment capturing how we arrive at this number, or get rid of the `reserve`.

Wajihulhassan replied:
I'll get rid of the `reserve` because I am not sure how memory saving it buys us.

bbannier commented on components/zeekopenbsm/src/openbsmconsumer.cpp:
```diff
@@ -0,0 +1,363 @@
+    tokens.push_back(tok);
+    bytesread += tok.len;
+  }
+  if (event_filter_list.find(event_id) == event_filter_list.end()) {
```
Doesn't this assume that the tokens we receive are ordered in a way where "interesting" tokens always preceed "uninteresting" ones? We would skip the rest of the function, even if we found some interesting ones.

Maybe I am misunderstanding what this does, but we also seem to reuse the `event_id` of the previous iteration if no case matches. Shouldn't we reset its value in a default case?

Wajihulhassan replied:
So it works like this: each event (e.g. socket connect) generates multiple token strings. Each generated token string has a type that is defined by `type.id`. HEADER is a must token type that appears only once per event. In this case, `event_id` will only be assigned once in the loop since there is only one token string with HEADER type.
I hope this anwsers your first question as well. Let me know if it is still unclear.

bbannier replied:
Thanks, that helps. Can we work on making the implementation more closely model what you explained? We could e.g., log a (fatal?) error if `au_fetch_tok` fails, move the storing of `tok` in `tokens` right after the fetch since they belong together, make `event_id` a `optional<u_int16_t>` and `assert(event_id)` before using it, and expand the comment for the early return with something capturing the guaranteed semantics.

Wajihulhassan replied:
I added your suggestion in the lastest commit.

bbannier commented on components/zeekopenbsm/src/openbsmconsumer.cpp:
```diff
@@ -0,0 +1,363 @@
+  if (audit_pipe == nullptr) {
+    throw Status::failure("The auditpipe couldn't be opened.");
+  }
+  event_filter_list.insert(AUE_CONNECT);
+  event_filter_list.insert(AUE_BIND);
```
Can we do this in the constructor instead? As written it can race with any reads of this variable.

bbannier commented on components/zeekopenbsm/src/openbsmconsumer.cpp:
```diff
@@ -0,0 +1,363 @@
+  }
+  event_filter_list.insert(AUE_CONNECT);
+  event_filter_list.insert(AUE_BIND);
+  fd_set fdset;
+  struct timeval timeout {};
+  timeout.tv_sec = 0;
+  timeout.tv_usec = 200000;
```
```suggestion

  fd_set fdset;

  timeval timeout;
  timeout.tv_sec = 0;
  timeout.tv_usec = 200'000;
```

bbannier commented on components/zeekopenbsm/src/openbsmconsumer.cpp:
```diff
@@ -0,0 +1,363 @@
+    }
+    if (rc < 0) {
+      if (errno != EINTR) {
+        return Status::failure("Auditpipe cannot be read");
```
Let's include `strerror(errno)` in the error message.

bbannier commented on components/zeekopenbsm/src/openbsmconsumer.cpp:
```diff
@@ -0,0 +1,363 @@
+      }
+      continue;
+    }
+    parseMessage();
```
We drop any error messages from this function on the floor. We should the very least log them, alternatively we could also transport them through (a modified) `event_list` to consumers.

Wajihulhassan replied:
Now `parseMessage` function logs any failure that happens in that function.

bbannier commented on components/zeekopenbsm/src/openbsmconsumer.cpp:
```diff
@@ -0,0 +1,363 @@
+    }
+    parseMessage();
+  }
+  return Status::success();
```
We never collect this return value. We could either transport it to the caller by having him pass a promise we set, or by just logging. In any case, this function could be `void`.

bbannier commented on components/zeekopenbsm/src/openbsmconsumer.cpp:
```diff
@@ -0,0 +1,363 @@
+    obj.reset(new OpenbsmConsumer(logger, configuration));
+    return Status::success();
+
+  } catch (const std::bad_alloc &) {
+    return Status::failure("Memory allocation failure");
```
Personally I find handling out-of-memory errors superfluous since we allocate in a lot of other places in the code where we never handle failures (e.g., we use `string`). I understand this is here for consistency with the rest of the code.

Wajihulhassan replied:
Yes, you are correct. I am doing this just to be consistent with the current code style.

bbannier commented on components/zeekopenbsm/tests/openbsmconsumer.cpp:
```diff
@@ -0,0 +1,103 @@
+#include "openbsmconsumer.h"
```
Could you include some tests here which exercise an actual `OpenbsmConsumer` instance?

Maybe it would make sense to change to a more end-to-end strategy here instead. For that you could change the constructor of `OpenbsmConsumer` so one can optionally pass in an alternative audit pipe path (defaulting to the "real" value). You could then inject events into a test pipe. That would also give us a chance to convince ourselves that we haven't introduced deadlocks, and to more exhaustively exercise the `extract*` functions where the current tests only cover a small subset.

Wajihulhassan replied:
I can add a test for e2e testing `OpenbsmConsumer` but that requires a lot of changes. I prefer to make a separate PR for that test. Further, as of now we only have one consumer and one producer running, there are fewer chances that we introduced some deadlocks.

bbannier replied:
We should really include a test of this functionality with this PR, currently we have almost no coverage. How did you test this?

Wajihulhassan replied:
The workflow of producer in openbsmconsumer.cpp is as follows:
`fetchRecordsFromAuditPipe` -> `parseRecordIntoTokens` -> `populateEventFromTokens` -> `extract*`.
Now, we have tests for `populateEventFromTokens` function which basically calls all the `extract*` methods to return `Event` object. Thus, our coverage is pretty good already.
To write a test that starts from `fetchRecordsFromAuditPipe` function is not that easy because you need to make sure in this new test that whatever you write to the pipe must be readable in `parseRecordIntoTokens`  using `au_read_rec` function. I cannot write some arbitrary text into the pipe and then read it using `au_read_rec`. 

Besides that, the grunt of work is done in `populateEventFromTokens` function, which we are already testing while `fetchRecordsFromAuditPipe` and `parseRecordIntoToken` functions just read records from the pipe and then tokenize them, respectively.

Wajihulhassan replied:
@bbannier To elaborate more on this, like all the other tests we have for Zeek Agent, we mock a log message from the underlying service (e.g., auditd, openbsm) _without_ running the service and then we have tests that checks the validity of our parsing functions. So I used the same pattern here. 
If you still think we should write a test that generates the log messages by running the actual service then could you give me some advice on how should I implement it.

bbannier replied:
> Besides that, the grunt of work is done in `populateEventFromTokens` function, which we are already testing ...

I had looked at coverage information from running the tests, and for some reason that reported no coverage where in the code there clearly is coverage. This seems good enough to me. Sorry for the hold-up because of that.

bbannier commented on tables/openbsm/src/openbsmservice.cpp:
```diff
@@ -0,0 +1,164 @@
+
+    return Status::success();
+
+  } catch (const std::bad_alloc &) {
+    return Status::failure("Memory allocation failure");
```
Ditto.

Wajihulhassan replied:
As I mentioned above, do we need explicit include here?

bbannier commented on tables/openbsm/src/openbsmservice.cpp:
```diff
@@ -0,0 +1,164 @@
+
+    return Status::success();
+
+  } catch (const std::bad_alloc &) {
+    return Status::failure("Memory allocation failure");
```
Ditto.

Wajihulhassan replied:
same as above.

Wajihulhassan committed:
Updating information regarding zeek-agent in readme (#70)

Wajihulhassan committed:
Apply suggestions from code review

Co-authored-by: Benjamin Bannier <bbannier@users.noreply.github.com>

Wajihulhassan commented on components/zeekopenbsm/src/openbsmconsumer.cpp:
```diff
@@ -0,0 +1,363 @@
+    : d(new PrivateData(logger, configuration)) {
+
+  // start fetching events from pipe in a separate thread
+  std::thread{&OpenbsmConsumer::fetchEventsFromPipe, this}.detach();
```
I did not use `join` because there was no reason for me to wait for consumer thread. As soon as the destructor for the producer is called, the consumer thread will be stopped. As of now, we don't have multiple Openbsm producers. We just have one Openbsm producer and one consumer; thus, leaking will never happen.

bbannier replied:
Not waiting here can obscure any issues in the shutdown of the threads (which might be introduced inadvertently a some later time). Let's make sure the thread actually has terminated by joining it.

Wajihulhassan replied:
Alright. That makes sense. I'll add `join` in the destructor.

Wajihulhassan commented on components/zeekopenbsm/src/openbsmconsumer.h:
```diff
@@ -0,0 +1,48 @@
+private:
+  struct PrivateData;
+  std::unique_ptr<PrivateData> d;
+  // audit pipe handler
+  FILE *audit_pipe{nullptr};
```
This was a bug. It should've been doxygen comment

Wajihulhassan commented on components/zeekopenbsm/src/openbsmconsumer.h:
```diff
@@ -0,0 +1,48 @@
+  /// \brief Constructor
+  OpenbsmConsumer(IZeekLogger &logger, IZeekConfiguration &configuration);
+
+  /// \brief extract header from openbsm token and populate event
+  static void extractHeader(Event &event, tokenstr_t tok);
+  /// \brief extract subject from openbsm token and populate event
+  static void extractSubject(Event &event, tokenstr_t tok);
+  /// \brief extract return from openbsm token and populate event
+  static void extractReturn(Event &event, tokenstr_t tok);
+  /// \brief extract socket-inet from openbsm token and populate event
+  static void extractSocketInet(Event &event, tokenstr_t tok);
```
No specific reason. Can this create some issue? What do you suggest?

bbannier replied:
Out parameters are harder to understand than simply returning a value since they e.g., force callers to default-construct a value which is only set up later. This can make it e.g., less obvious whether a value is set on all paths.

I'd suggest changing all the functions here to take a `tokenstr_t` and to return an `Event` by value. `populateSocketEvent` take a `const vector<tokenstr_t>&` should return an `Event` as well (the status is not needed).

Wajihulhassan replied:
In the latest commit, these functions return `Status` object to show if the extraction was successful or not so that we can log in a failure scenario. Let me know what you think.

Wajihulhassan commented on tables/openbsm/include/zeek/openbsmservicefactory.h:
```diff
@@ -0,0 +1,18 @@
+///                         are registered
+/// \param configuration An initialized configuration object
+/// \param logger An initialized logger object
+Status registerOpenbsmServiceFactory(IZeekServiceManager &service_manager,
```
Do we explicitly need to add `zeek/status.h` here if already included file `zeek/izeekconfiguration.h` has `zeek/status.h` in its includes?

bbannier replied:
I'd argue we should include it to make this file independent of other header files (even ones it includes). https://google.github.io/styleguide/cppguide.html#Self_contained_Headers

Wajihulhassan commented on tables/openbsm/include/zeek/openbsmservicefactory.h:
```diff
@@ -0,0 +1,18 @@
+/// \param configuration An initialized configuration object
+/// \param logger An initialized logger object
+Status registerOpenbsmServiceFactory(IZeekServiceManager &service_manager,
+                                     IVirtualDatabase &virtual_database,
```
same as above.

Wajihulhassan commented on tables/openbsm/include/zeek/openbsmservicefactory.h:
```diff
@@ -0,0 +1,18 @@
+Status registerOpenbsmServiceFactory(IZeekServiceManager &service_manager,
+                                     IVirtualDatabase &virtual_database,
+                                     IZeekConfiguration &configuration,
+                                     IZeekLogger &logger);
```
same as above.

Commit:
as per benjamin comments: I fixed issue with returned statuses and handled unix sockets

Commit:
updating variable and function names; adding headers

bbannier commented on components/zeekopenbsm/src/openbsmconsumer.cpp:
```diff
@@ -0,0 +1,361 @@
+
+  if (subscribed_event_ids.find(event_id.value()) ==
+      subscribed_event_ids.end()) {
+    // we will not store current record in the event_list
```
Let's document _why_ instead of _how_, e.g.,
```suggestion
    // Skip events not subscribed to.
```

bbannier commented on components/zeekopenbsm/src/openbsm_utils.cpp:
```diff
@@ -0,0 +1,33 @@
+  if (sock.family == 2) {
+    struct in_addr ipv4 {};
+    ipv4.s_addr = static_cast<in_addr_t>(*sock.addr);
+    ip_addr = std::string(inet_ntop(AF_INET, &ipv4, ip_str, INET6_ADDRSTRLEN));
+    return Status::success();
+  } else if (sock.family == 26) {
+    struct in6_addr ipv6 {};
+    memcpy(&ipv6, sock.addr, sizeof(ipv6));
+    ip_addr = std::string(inet_ntop(AF_INET6, &ipv6, ip_str, INET6_ADDRSTRLEN));
+    return Status::success();
+  }
+  return Status::failure("Cannot parse IP address from given token due to "
+                         "unhandled socket family");
+}
+
+Status getPathFromPid(int pid, std::string &path) {
+  char pathbuf[PROC_PIDPATHINFO_MAXSIZE] = {0};
+
+  int ret = proc_pidpath(pid, pathbuf, sizeof(pathbuf));
+  if (ret > 0) {
+    path = std::string(pathbuf);
```
Small nit since I see it here:
```suggestion
    ip_addr = inet_ntop(AF_INET, &ipv4, ip_str, INET6_ADDRSTRLEN);
    return Status::success();
  } else if (sock.family == 26) {
    struct in6_addr ipv6 {};
    memcpy(&ipv6, sock.addr, sizeof(ipv6));
    ip_addr = inet_ntop(AF_INET6, &ipv6, ip_str, INET6_ADDRSTRLEN);
    return Status::success();
  }
  return Status::failure("Cannot parse IP address from given token due to "
                         "unhandled socket family");
}

Status getPathFromPid(int pid, std::string &path) {
  char pathbuf[PROC_PIDPATHINFO_MAXSIZE] = {0};

  int ret = proc_pidpath(pid, pathbuf, sizeof(pathbuf));
  if (ret > 0) {
    path = pathbuf;
```

bbannier reviewed (commented):
Looks good to me but for two additional nits and the outstanding test.

Wajihulhassan committed:
Apply suggestions from code review

Co-authored-by: Benjamin Bannier <bbannier@users.noreply.github.com>

bbannier committed:
Make sure libcrypto is linked in after libssl (#72)

* Make dependency of libcrypto on libssl explicit

libcrypto depends on libssl. With this patch we now explicitly express
that dependency.

Closes #71.

Wajihulhassan committed:
Adding file events table to Linux platform and standardize variable/column names across tables (#68)

This PR basically adds/updates following three big things:

1. Adds a file events table fileeventstableplugin.cpp for Linux platform. Currently, it handles creat, open, and openat syscalls. I have also added tests that generate file event logs and then check table rows generated by zeek-agent for data correctness.

2. Updates column names for all three tables (process, socket, and file) so that those column names match the key names that appear in the underlying auditd logs. For example, it replaces "parent" column name with "ppid" because auditd log uses "ppid" key name to represent parent's pid. I also updated the variable names in processeventstableplugin.cpp, socketeventstableplugin.cpp, and fileeventstableplugin.cpp files so that they are consistent and match across tables.

3. Incorporates requested changes on macOS event tables in PR#67 that are also applicable here to Linux event tables.

bbannier commented:
Looks good to me and ready for merge once CI passes.

bbannier commented:
@Wajihulhassan, can you rebase this on the current `master` and check CI? Like I wrote, if CI passes, you could then go ahead and merge this.
